### PR TITLE
Property based testing blog post

### PR DESCRIPTION
**Pull request by selaux**

### CODE REVIEW

File: _posts/2017-07-26-property-based-testing-in-javascript.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+```
+
+When a value for this generator is built, we basically get a function that checks whether the hash of the value starts
+with `i`, where `i` is a value drawn from the range from `0` to `9`. This function is deterministic, so we can use it
+as the input for our property test. This concept of generator transformation exists in all of the mentioned frameworks
+(although it is called `map` in the others).
+
+## Conclusion
+
+Property-based testing is a cool and useful addition to your test suite. We showed you how to build your own property-based
```
Comment (HeeL) [inline _posts/2017-07-26-property-based-testing-in-javascript.md]: Replace `we showed you` with i.e. `we saw`, because it's basically you, who showed us ;)

File: _posts/2017-07-26-property-based-testing-in-javascript.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+[Array.prototype.filter][filter].
+You can find the full example, where you can actually run the tests, [here][examples].
+
+```javascript
+function filter(fn, array) {
+    return array.filter(fn);
+}
+```
+
+As the first step we need to we have to define properties for this function. I picked some, but there might be additional
```
Comment (HeeL) [inline _posts/2017-07-26-property-based-testing-in-javascript.md]: Either `we need to` or `we have to`

**Comment by HeeL** [comment]
Very nice writing

**Comment by gilligan** [comment]
It somehow feels quite distracting that the font sizes are changing like that. But that most probably doesn't have anything to do with your post but with our global css or whatever? 

If I am the only one noticing/disliking this then I'm sure it is a non-issue though.

File: _posts/2017-07-26-property-based-testing-in-javascript.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+show_avatar : true
+read_time : 10
+show_related_posts: false
+---
+
+As developers we are always looking for new ways to ensure that our code works as we would expected. One way to
+achieve this, which is not as commonly used as for example unit tests, is property-based testing. Property-based testing
+is a testing method based on (pseudo-)randomly generated input. It can be used at any level of the testing pyramid to
+validate some properties of algorithms or systems we built and is best used in addition to other tests, which test
+specific instances of inputs.
```
Comment (gilligan) [inline _posts/2017-07-26-property-based-testing-in-javascript.md]: You are already drawing some conclusions in the first paragraph of your post without validating them. Why can they be added "at any level of the testing pyramid" and why are they "best used in addition" ? Why can't they replace unit tests entirely? (I am not advocating that-  just asking the question).

File: _posts/2017-07-26-property-based-testing-in-javascript.md [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+The test runner will then roughly do the following when executing your test:
+
+1. Ideally it will seed your generator function with a random seed. This is necessary to make the results reproducible
+1. It will feed the next output of your generator function into the property predicate
+1. It will check whether the output of your unit-under-test passes the property predicate
+   - If false: fail test and print the input that lead to non-compliance with the property description
+   - If true: continue with 2. (until some number of iterations is reached)
+
+A simple example for a property based test would be the following property based test for the plus function (for
+simplicities sake I left out the seed part of the generator function). This property-based test tests whether the
```
Comment (gilligan) [inline _posts/2017-07-26-property-based-testing-in-javascript.md]: for simplicity's sake

File: _posts/2017-07-26-property-based-testing-in-javascript.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+- **Property predicate**: `const prop = (in1, in2) => add(in1, in2) >= 0`
+
+The advantage of this type of testing is that your function will be tested with way more input values compared to testing
+with regular unit tests. This allows you to find edge cases where your unit-under-test does not behave as you would expect
+it to. This is quite similar to Fuzzing in the IT-Security community, where the goal is to find security issues or memory
+leaks by providing (invalid) random data.
+
+A disadvantage is that, depending on the unit-under-test, the properties can be quite hard to define. For the simple
+plus example from before it was quite easy, but different algorithms have different properties and they are often very unique
+to the unit-under-test. Another drawback is that property-based testing does not guaranteed to find any existing issues with any
```
Comment (gilligan) [inline _posts/2017-07-26-property-based-testing-in-javascript.md]: I'm not quite sure I follow. First of: are you talking about pros and cons of property based tests in contrast to unit tests? Furthermore: 

> Another drawback is that property-based testing does not guaranteed to find any existing issues with any

How do unit tests give you any guarantee that you are finding problems with the cases that you came up with?

File: _posts/2017-07-26-property-based-testing-in-javascript.md [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+test-run. The goal is rather to find them eventually by generating enough random data.
+
+## Property-Based Testing in JavaScript
+
+In other languages like Haskell ([QuickCheck][quickcheck]) and 
+Clojure ([test.check][test.check]) there is a de-facto standard for property based testing.
```
Comment (gilligan) [inline _posts/2017-07-26-property-based-testing-in-javascript.md]: I guess what you mean is that in Haskell and Clojure property based tests are commonly used? The use of "standard" is a bit confusing. There is no "standard" ...

File: _posts/2017-07-26-property-based-testing-in-javascript.md [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+In the JavaScript world this is not yet the case. I found few frameworks that are actively developed and have a
+substantial amount of users (> 1000 downloads per month on npm).
```
Comment (gilligan) [inline _posts/2017-07-26-property-based-testing-in-javascript.md]: It would be nice if you could also list some projects that are actually using any of those libraries. For example Ramda: https://github.com/ramda/ramda/search?utf8=%E2%9C%93&q=jsverify&type= or sanctuary https://github.com/sanctuary-js/sanctuary/search?utf8=%E2%9C%93&q=jsverify&type=

**Comment by gilligan** [comment]
I am missing some conclusion that adds to the statement in your first paragraph about property based tests being good in *addition* to unit tests - why though? Because unit tests are a good way to document how to *use* something whereas property based tests are a good way to establish that the code behaves according to certain rules. 

The one cannot easily be deduced from the other.

File: _posts/2017-07-26-property-based-testing-in-javascript.md (hunk 2 above)
Comment (selaux) [inline _posts/2017-07-26-property-based-testing-in-javascript.md]: Done.

File: _posts/2017-07-26-property-based-testing-in-javascript.md (hunk 1 above)
Comment (selaux) [inline _posts/2017-07-26-property-based-testing-in-javascript.md]: Done.

File: _posts/2017-07-26-property-based-testing-in-javascript.md (hunk 3 above)
Comment (selaux) [inline _posts/2017-07-26-property-based-testing-in-javascript.md]: True. I moved this a little further down and explained it a little more.

File: _posts/2017-07-26-property-based-testing-in-javascript.md (hunk 4 above)
Comment (selaux) [inline _posts/2017-07-26-property-based-testing-in-javascript.md]: Thx, fixed.

**Comment by gilligan** [comment]
Oh by the way, @seelaux what i totally forgot to mention was: awesome job writing about property based tests in the first place! 👍👍👍

File: _posts/2017-07-26-property-based-testing-in-javascript.md (hunk 5 above)
Comment (selaux) [inline _posts/2017-07-26-property-based-testing-in-javascript.md]: Yeah, it is not really drawbacks, but rather things to be aware of. I rephrased the paragraph and included the conclusions from above. I hope this makes things more clear

File: _posts/2017-07-26-property-based-testing-in-javascript.md (hunk 6 above)
Comment (selaux) [inline _posts/2017-07-26-property-based-testing-in-javascript.md]: 👍 Wording adapted.

File: _posts/2017-07-26-property-based-testing-in-javascript.md (hunk 7 above)
Comment (selaux) [inline _posts/2017-07-26-property-based-testing-in-javascript.md]: I think this would make this list kind of complicated. Maybe adding some examples at the end makes sense though...

**Comment by selaux** [comment]
Yeah these different sizes for lists don't really make sense to me as well. Additionally the width of the page is very small on desktop. Why do I have to read like i'm on a smartphone. But we can fix that in another PR.
